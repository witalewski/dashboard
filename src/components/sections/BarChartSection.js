import React from 'react';
import {
  ComposedChart,
  Line,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
} from 'recharts';
import { chartColors } from '../../global/styleConstants';
import { Section } from '../Section';

const processData = (labelMap, data) =>
  data.map(item =>
    Object.keys(item).reduce(
      (acc, key) => ({ ...acc, [labelMap[key]]: item[key] }),
      {}
    )
  );

const getLowerRangeSeriesKeys = (labelMap, processedData) => {
  const sortedSeries = Object.values(labelMap)
    .filter(e => e !== 'label')
    .map(key => ({
      key: key,
      avgValue:
        processedData.reduce((acc, item) => acc + item[key], 0) /
        processedData.length,
    }))
    .sort((a, b) => a.avgValue - b.avgValue);
  const halfOfSortedSeriesLength = Math.floor(sortedSeries.length / 2);
  const lowerRangeSeries = sortedSeries.slice(0, halfOfSortedSeriesLength);
  if (sortedSeries.length % 2 !== 0) {
    let middleItem = sortedSeries[halfOfSortedSeriesLength];
    if (
      Math.abs(
        middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength - 1]
      ) <
      Math.abs(middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength + 1])
    ) {
      lowerRangeSeries.push(middleItem);
    }
  }
  return lowerRangeSeries.map(item => item.key);
};

export const BarChartSection = ({ className, data, title, labelMap }) => {
  const processedData = processData(labelMap, data);
  const lowerRangeSeriesKeys = getLowerRangeSeriesKeys(labelMap, processedData);
  return (
    <Section title={title} className={className}>
      <ComposedChart
        width={600}
        height={400}
        data={processedData}
        margin={{ top: 20, right: 20, bottom: 20, left: 20 }}
      >
        <CartesianGrid stroke="#f5f5f5" />
        <XAxis dataKey={Object.keys(data[0])[0]} />
        <YAxis yAxisId="lower" />
        <YAxis yAxisId="upper" orientation="right" />
        <Tooltip />
        <Legend />
        {Object.keys(data[0])
          .filter(key => key.match(/bar/))
          .map((key, i) => (
            <Bar
              key={labelMap[key]}
              dataKey={labelMap[key]}
              barSize={20}
              fill={chartColors[i % chartColors.length]}
              yAxisId={
                lowerRangeSeriesKeys.indexOf(labelMap[key]) > -1
                  ? 'lower'
                  : 'upper'
              }
            />
          ))}
        {Object.keys(data[0])
          .filter(key => key.match(/line/))
          .map((key, i) => (
            <Line
              key={labelMap[key]}
              type="monotone"
              dataKey={labelMap[key]}
              stroke={chartColors[i % chartColors.length]}
              yAxisId={
                lowerRangeSeriesKeys.indexOf(labelMap[key]) > -1
                  ? 'lower'
                  : 'upper'
              }
            />
          ))}
      </ComposedChart>
    </Section>
  );
};
