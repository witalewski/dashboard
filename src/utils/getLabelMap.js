export const getLabelMap = source => ({
  label: 'label',
  ...Object.keys(source)
    .filter(key => key.match(/-label/))
    .reduce((acc, key) => ({ ...acc, [key.split('-')[0]]: source[key] }), {}),
});
