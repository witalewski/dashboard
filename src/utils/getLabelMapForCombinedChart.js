export const getLabelMapForCombinedChart = source => ({
  label: 'label',
  ...Object.keys(source)
    .filter(key => key.match(/-bar|-label/))
    .reduce((acc, key) => ({ ...acc, [key.split('-')[0]]: source[key] }), {}),
});
