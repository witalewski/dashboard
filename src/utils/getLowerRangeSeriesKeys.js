export const getLowerRangeSeriesKeys = (labelMap, processedData) => {
  const sortedSeries = Object.values(labelMap)
    .filter(e => e !== 'label')
    .map(key => ({
      key: key,
      avgValue:
        processedData.reduce((acc, item) => acc + item[key], 0) /
        processedData.length,
    }))
    .sort((a, b) => a.avgValue - b.avgValue);
  const halfOfSortedSeriesLength = Math.floor(sortedSeries.length / 2);
  const lowerRangeSeries = sortedSeries.slice(0, halfOfSortedSeriesLength);
  if (sortedSeries.length % 2 !== 0) {
    let middleItem = sortedSeries[halfOfSortedSeriesLength];
    if (
      Math.abs(
        middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength - 1]
      ) <
      Math.abs(middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength + 1])
    ) {
      lowerRangeSeries.push(middleItem);
    }
  }
  return lowerRangeSeries.map(item => item.key);
};
