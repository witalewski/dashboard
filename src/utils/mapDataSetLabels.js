export const mapDataSetLabels = (labelMap, data) =>
  data.map(item =>
    Object.keys(item).reduce(
      (acc, key) => ({ ...acc, [labelMap[key]]: item[key] }),
      {}
    )
  );
