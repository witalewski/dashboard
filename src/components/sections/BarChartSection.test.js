import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import mockChartData from '../../fixtures/mockChartData.json';
import { BarChartSection } from './BarChartSection';
import { getLabelMapForCombinedChart } from '../../utils/getLabelMapForCombinedChart';

describe('BarChartSection', () => {
  it('matches snapshot with mock data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <BarChartSection
        title="Monthly Sales"
        data={mockChartData.monthlySales.data}
        dataTransform={e => e}
        labelMap={getLabelMapForCombinedChart(mockChartData.monthlySales)}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
