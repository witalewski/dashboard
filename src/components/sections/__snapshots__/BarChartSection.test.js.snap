// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChartSection matches snapshot with mock data 1`] = `
<Section
  title="Monthly Sales"
>
  <ComposedChart
    barCategoryGap="10%"
    barGap={4}
    data={
      Array [
        Object {
          "Event": 120.65,
          "Pass": 6562.67,
          "Single/Rental": null,
          "Subscription": null,
          "Transactions": 12455,
          "label": "December",
        },
        Object {
          "Event": 243.64,
          "Pass": 7165.2,
          "Single/Rental": null,
          "Subscription": null,
          "Transactions": 13468,
          "label": "January",
        },
        Object {
          "Event": 345.54,
          "Pass": 46744.11,
          "Single/Rental": 12.4,
          "Subscription": null,
          "Transactions": 58446,
          "label": "February",
        },
      ]
    }
    height={400}
    layout="horizontal"
    margin={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
    width={600}
  >
    <CartesianGrid
      fill="none"
      horizontal={true}
      horizontalFill={Array []}
      horizontalPoints={Array []}
      stroke="#f5f5f5"
      vertical={true}
      verticalFill={Array []}
      verticalPoints={Array []}
    />
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="label"
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={30}
      hide={false}
      mirror={false}
      orientation="bottom"
      padding={
        Object {
          "left": 0,
          "right": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="category"
      width={0}
      xAxisId={0}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={0}
      hide={false}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="number"
      width={60}
      yAxisId="lower"
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={0}
      hide={false}
      mirror={false}
      orientation="right"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="number"
      width={60}
      yAxisId="upper"
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
    <Legend
      align="center"
      iconSize={14}
      layout="horizontal"
      verticalAlign="bottom"
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      barSize={20}
      data={Array []}
      dataKey="Event"
      fill="#5db990"
      hide={false}
      isAnimationActive={true}
      layout="vertical"
      legendType="rect"
      minPointSize={0}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      xAxisId={0}
      yAxisId="upper"
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      barSize={20}
      data={Array []}
      dataKey="Single/Rental"
      fill="#558ca0"
      hide={false}
      isAnimationActive={true}
      layout="vertical"
      legendType="rect"
      minPointSize={0}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      xAxisId={0}
      yAxisId="lower"
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      barSize={20}
      data={Array []}
      dataKey="Pass"
      fill="#864ba5"
      hide={false}
      isAnimationActive={true}
      layout="vertical"
      legendType="rect"
      minPointSize={0}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      xAxisId={0}
      yAxisId="upper"
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      barSize={20}
      data={Array []}
      dataKey="Subscription"
      fill="#829ac0"
      hide={false}
      isAnimationActive={true}
      layout="vertical"
      legendType="rect"
      minPointSize={0}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      xAxisId={0}
      yAxisId="lower"
    />
    <Line
      activeDot={true}
      animateNewValues={true}
      animationBegin={0}
      animationDuration={1500}
      animationEasing="ease"
      connectNulls={false}
      dataKey="Transactions"
      dot={true}
      fill="#fff"
      hide={false}
      isAnimationActive={true}
      legendType="line"
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      points={Array []}
      stroke="#5db990"
      strokeWidth={1}
      type="monotone"
      xAxisId={0}
      yAxisId="upper"
    />
  </ComposedChart>
</Section>
`;
